recorder/src/main/cpp/profile_writer.cc

```diff
@@ -166,43 +169,91 @@ void ProfileSerializingWriter::record(const JVMPI_CallTrace &trace, ThreadBucket
     if (snipped) s_c_frame_snipped.inc();
     ss->set_snipped(snipped);
 
-    if (trace.num_frames <= 0) {
+    if (trace.error != BacktraceError::Fkp_no_error) {
         s_m_stack_sample_err.mark();
-        return;
     }
-    if (ctx_len == 0) {
-        ss->add_trace_id(NOCTX_ID);
+    
+    if (default_ctx) {
+        assert(ctx_len == 0);
+        ss->add_trace_id(DEFAULT_CTX_ID);
+    } else if (ctx_len == 0) {
+        ss->add_trace_id(UNKNOWN_CTX_ID);
     }
 
+    auto bt_type = trace.type;
     for (auto i = 0; i < Util::min(static_cast<TruncationCap>(trace.num_frames), trunc_thresholds.cpu_samples_max_stack_sz); i++) {
         auto f = ss->add_frame();
-        auto& jvmpi_cf = trace.frames[i];
-        //find method
-        auto mth_id = jvmpi_cf.method_id;
-        if (known_methods.count(reinterpret_cast<MthId>(mth_id)) == 0) {
-            if (fir(mth_id, jvmti, *this)) {
-                s_c_new_mthd_info.inc();
-            } else {
-                recordNewMethod(mth_id, "?", "?", "?", "?");
-            }
-            s_c_total_mthd_info.inc();
+        switch (bt_type) {
+        case BacktraceType::Java:
+            fill_frame(trace.frames[i].jvmpi_frame, f);
+            break;
+        case BacktraceType::Native:
+            fill_frame(trace.frames[i].native_frame, f);
+            break;
+        default:
+            assert(false);
         }
-        //end find method
-        f->set_method_id(reinterpret_cast<std::int64_t>(mth_id));
-        f->set_bci(jvmpi_cf.lineno);//turns out its actually BCI
-        auto line_no = lnr(jvmpi_cf.lineno, mth_id, jvmti);
-        if (line_no < 0) s_c_bad_lineno.inc();
-        f->set_line_no(line_no);
     }
 
     s_m_cpu_sample_add.mark();
 }
 
-void ProfileSerializingWriter::recordNewMethod(const jmethodID method_id, const char *file_name, const char *class_name, const char *method_name, const char *method_signature) {
-    known_methods.insert(reinterpret_cast<MthId>(method_id));
+void ProfileSerializingWriter::fill_frame(const JVMPI_CallFrame& jvmpi_cf, recording::Frame* f) {
+    auto mth_id = jvmpi_cf.method_id;
+    MthId my_id;
+    auto it = known_methods.find(reinterpret_cast<MthId>(mth_id));
```

Comment:
Is doing reinterpret_cast<MthId> a good thing here? Wouldn't a static_cast suffice.